File: fixtures/scale/kotlin/large/src/cryptobench/ciphers/rc4/Rc4Cipher.kt
package cryptobench.ciphers.rc4

import cryptobench.core.Cipher
import cryptobench.util.Bytes
import cryptobench.util.Hex

class Rc4Cipher(key: Rc4Key) : Cipher {
    private val keyBytes = Bytes.of(key.secret)

    override fun name(): String = "rc4"

    override fun encrypt(plaintext: String): String = Hex.encode(stream(Bytes.of(plaintext)))

    override fun decrypt(ciphertext: String): String = Bytes.toText(stream(Hex.decode(ciphertext)))

    private fun stream(data: ByteArray): ByteArray {
        val s = IntArray(256) { it }
        var j = 0
        for (i in 0 until 256) {
            j = (j + s[i] + (keyBytes[i % keyBytes.size].toInt() and 0xFF)) and 0xFF
            s[i] = s[j].also { s[j] = s[i] }
        }
        val out = ByteArray(data.size)
        var x = 0
        var y = 0
        for (n in data.indices) {
            x = (x + 1) and 0xFF
            y = (y + s[x]) and 0xFF
            s[x] = s[y].also { s[y] = s[x] }
            out[n] = (data[n].toInt() xor s[(s[x] + s[y]) and 0xFF]).toByte()
        }
        return out
    }
}

File: fixtures/scale/dart/large/lib/ciphers/rc4/rc4_cipher.dart
import 'dart:typed_data';

import '../../core/cipher.dart';
import '../../util/bytes.dart';
import '../../util/hex.dart';
import 'rc4_key.dart';

class Rc4Cipher implements Cipher {
  final Uint8List _keyBytes;

  Rc4Cipher(Rc4Key key) : _keyBytes = bytesOf(key.secret);

  @override
  String name() => 'rc4';

  @override
  String encrypt(String plaintext) => hexEncode(_stream(bytesOf(plaintext)));

  @override
  String decrypt(String ciphertext) => bytesToText(_stream(hexDecode(ciphertext)));

  Uint8List _stream(Uint8List data) {
    final s = List<int>.generate(256, (i) => i);
    var j = 0;
    for (var i = 0; i < 256; i++) {
      j = (j + s[i] + _keyBytes[i % _keyBytes.length]) & 0xFF;
      final tmp = s[i];
      s[i] = s[j];
      s[j] = tmp;
    }
    final out = Uint8List(data.length);
    var x = 0, y = 0;
    for (var n = 0; n < data.length; n++) {
      x = (x + 1) & 0xFF;
      y = (y + s[x]) & 0xFF;
      final tmp = s[x];
      s[x] = s[y];
      s[y] = tmp;
      out[n] = data[n] ^ s[(s[x] + s[y]) & 0xFF];
    }
    return out;
  }
}
